src/libraries/System.Private.CoreLib/src/System/Collections/Generic/ReferenceEqualityComparer.cs

```diff
@@ -0,0 +1,53 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Runtime.CompilerServices;
+
+namespace System.Collections.Generic
+{
+    /// <summary>
+    /// An <see cref="IEqualityComparer{Object}"/> that uses reference equality (<see cref="object.ReferenceEquals(object?, object?)"/>)
+    /// instead of value equality (<see cref="object.Equals(object?)"/>) when comparing two object instances.
+    /// </summary>
+    /// <remarks>
+    /// The <see cref="ReferenceEqualityComparer"/> type cannot be instantiated. Instead, use the <see cref="Instance"/> property
+    /// to access the singleton instance of this type.
+    /// </remarks>
+    public sealed class ReferenceEqualityComparer : IEqualityComparer<object>, IEqualityComparer
+    {
+        private ReferenceEqualityComparer() { }
+
+        /// <summary>
+        /// Gets the singleton <see cref="ReferenceEqualityComparer"/> instance.
+        /// </summary>
+        public static ReferenceEqualityComparer Instance { get; } = new ReferenceEqualityComparer();
+
+        /// <summary>
+        /// Determines whether two object references refer to the same object instance.
+        /// </summary>
+        /// <param name="x">The first object to compare.</param>
+        /// <param name="y">The second object to compare.</param>
+        /// <returns>
+        /// <see langword="true"/> if both <paramref name="x"/> and <paramref name="y"/> refer to the same object instance
+        /// or if both are <see langword="null"/>; otherwise, <see langword="false"/>.
+        /// </returns>
+        /// <remarks>
+        /// This API is a wrapper around <see cref="object.ReferenceEquals(object?, object?)"/>.
+        /// It is not necessarily equivalent to calling <see cref="object.Equals(object?, object?)"/>.
+        /// </remarks>
+        public new bool Equals(object? x, object? y) => ReferenceEquals(x, y);
+
+        /// <summary>
+        /// Returns a hash code for the specified object. The returned hash code is based on the object
+        /// identity, not on the contents of the object.
+        /// </summary>
+        /// <param name="obj">The object for which to retrieve the hash code.</param>
+        /// <returns>A hash code for the identity of <paramref name="obj"/>.</returns>
+        /// <remarks>
+        /// This API is a wrapper around <see cref="RuntimeHelpers.GetHashCode(object)"/>.
+        /// It is not necessarily equivalent to calling <see cref="object.GetHashCode()"/>.
+        /// </remarks>
+        public int GetHashCode(object obj) => RuntimeHelpers.GetHashCode(obj);
```

Comment:
shouldn't this be `public int GetHashCode(object? obj)`?  `null` is certainly supported there (and also through EqualityComparer<T>)

Reply:
Yes. The base interface advertises object, but the derived type can advertise that it's more accepting.

Reply:
Whew, I was regretting not thinking to ask this with the initial proposal. The ability to do `new HashSet<string?>(ReferenceEqualityComparer.Instance) { null }` is going to be important to me.

Reply:
It wouldn't affect that. HashSet won't use the comparer's GetHashCode with a null value.

Reply:
I didn't know that, thanks! I think I have be calling IEqualityComparer<T>.GetHashCode without specifically guarding against `null` for years, so I have some catching up to do. :)

Reply:
When we reviewed the API we had originally decided _against_ allowing null inputs to `GetHashCode(object)`, as both `IEqualityComparer<T>` and `RuntimeHelpers.GetHashCode` disallow null inputs per their nullability annotations. (In practice `RuntimeHelpers.GetHashCode` does allow null inputs and normalizes the hash code to 0.)

`IEqualityComparer<T>.GetHashCode` goes further and [explicitly documents](https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.iequalitycomparer-1.gethashcode) that null is disallowed and that this will result in an exception, though in practice it depends on the implementation. For example, `StringComparer.Ordinal.GetHashCode` will throw an exception on null, but `EqualityComparer<T>.Default.GetHashCode` will return 0.

If there's really a desire to support `GetHashCode` accepting null (why?) we can provide the annotation for it. It just seems strange to do so.

Reply:
@GrabYourPitchforks The current definition of ReferenceEqualityComparer in this PR causes this problem:

```cs
void M<T>()
{
    // CS8620 Argument of type 'ReferenceEqualityComparer' cannot be used for parameter
    // 'comparer' of type 'IEqualityComparer<string?>' in 'HashSet<string?>.HashSet(
    // IEqualityComparer<string?>? comparer)' due to differences in the nullability of
    // reference types.  ↓
    new HashSet<string?>(ReferenceEqualityComparer.Instance);
}
```

Reply:
> In practice RuntimeHelpers.GetHashCode does allow null inputs and normalizes the hash code to 0

The annotation should be fixed to match it.

Reply:
Yes, we should fix the annotation on RuntimeHelpers.

As for `IEqualityComparer<T>.GetHashCode`, it documents that a caller needs to expect an exception for null is possible; it doesn't guarantee an exception, exactly because as you say implementations can and do make better guarantees. This implementation allows null, and as a public type we have the ability to advertise that through its annotations; there's little downside to doing so.